app/controllers/api/v0/versions_controller.rb

```diff
@@ -25,6 +25,23 @@ def show
     }
   end
 
+  def raw
+    @version ||= version_collection.find(params[:id])
+
+    expires_in 1.year, :public => true
+
+    if Archiver.public_archive_url?(@version.uri)
+      redirect_to @version.uri, status: 301 and return
+    else
+      upstream = Archiver.get_file_from_uri(@version.uri)
```

Comment:
It looks like there's some magic happening here. Actually implementing version_hash as the ETag value is looking more complicated than I first thought it would be. Unless I'm missing something, it looks like we'd either need to implement our own cache_key method for versions, or else manually `response.set_header` and do our own conditional response handling. It does look like there are already conditional responses based on automagical weak Etags.

Reply:
> It does look like there are already conditional responses based on automagical weak Etags

I guess that doesn't seem surprising. But it also seems like the right etag (and actually, maybe it should be strong?) in this case is our `version_hash`, so we’d rather have that than what Rails automatically gins up for us.

All that said, it looks like it’s ultimately setting the etag to an MD5 hex digest of whatever you pass in, which is a little pointless in our case, but probably not harmful. As long as it’s deterministic, all good?

Reply:
(I should say, MD5 hex digest of the explicit etag we pass in + last modified date + anything extra the controller specified should be in all etags.)

Reply:
Way more special and less direct than I’d realized whenever I’d used `stale?` before!

Reply:
> As long as it’s deterministic, all good?

Works for me!